x-pack/platform/plugins/shared/fleet/public/applications/fleet/sections/agents/components/agent_health.tsx

```diff
@@ -42,9 +42,11 @@ type Props = EuiBadgeProps & {
 
 function getStatusComponent({
   status,
+  inFailedUpgradeState,
   ...restOfProps
 }: {
   status: Agent['status'];
+  inFailedUpgradeState: boolean;
```

Comment:
it feels a bit odd to introduce a new prop just for this specific sub-state.. I wonder if we can just accept the agent upgrade details information like this:
```suggestion
  status: Agent['status'];
  upgradeDetails?: Agent['upgrade_details'];
```

and then use `isAgentInFailedUpgradeState()` directly in this component to derive the badge. WDYT?

then this component could show other upgrade sub-states in the future if we choose to (I see 9 of them in `AgentUpgradeStateType`!)

Reply:
I agree, using something more extensible rather than a specific prop makes more sense in this case. However, that `isAgentInFailedUpgradeState()` function takes an `Agent` rather than `upgrade_details`, so maybe it makes sense to simplify further and just pass the entire Agent into this function and then we can access the `status` and `upgrade_details` inside as needed. WDYT?

Reply:
> that isAgentInFailedUpgradeState() function takes an Agent rather than upgrade_details

this is true, but it is only using `upgrade_details` in its logic so can be easily refactored

Reply:
> this is true, but it is only using `upgrade_details` in its logic so can be easily refactored

Should we prefer refactoring that function then over my latest changes (passing in the Agent and using a union for the props type)?

Reply:
IMO `status`+`upgradeDetails` makes more sense as props for this component than `agent`+`previousStatus`